types/validator_set_test.go

```diff
@@ -599,3 +599,217 @@ func TestValidatorSetVerifyCommit(t *testing.T) {
 	err = vset.VerifyCommit(chainID, blockID, height, commit)
 	assert.Nil(t, err)
 }
+
+
+func TestValidatorUpdates(t *testing.T) {
+
+	v1 := newValidator([]byte("v1"), 100)
+	v2 := newValidator([]byte("v2"), 100)
+	valList := []*Validator{v1, v2}
+	vals := NewValidatorSet(valList)
+	assert.Nil(t, verifyValidatorSet(vals))
+
+	// noop
+	valList = []*Validator{}
+	valsc := vals.Copy()
+	assert.Nil(t, vals.UpdateWithChangeSet(valList))
+	assert.Nil(t, verifyValidatorSetEquiv(vals, valsc))
+
+	// Verify duplicate updates are caught and vals is not modified
+	v1 = newValidator([]byte("v3"), 10)
+	v2 = newValidator([]byte("v2"), 200)
+	valList = []*Validator{v1, v2, v1}
+	valsc = vals.Copy()
+	assert.NotNil(t, vals.UpdateWithChangeSet(valList))
+	assert.Nil(t, verifyValidatorSetEquiv(vals, valsc))
+
+	// Verify overflows are caught and vals is not modified
+	v1 = newValidator([]byte("v3"), 10)
+	v2 = newValidator([]byte("v2"), math.MaxInt64)
+	valList = []*Validator{v1, v2}
+	valsc = vals.Copy()
+	assert.NotNil(t, vals.UpdateWithChangeSet(valList))
+	assert.Nil(t, verifyValidatorSetEquiv(vals, valsc))
+
+	// Verify duplicate removes are caught and vals is not modified
+	v2 = newValidator([]byte("v2"), 0)
+	valList = []*Validator{v2, v2}
+	valsc = vals.Copy()
+	assert.NotNil(t, vals.UpdateWithChangeSet(valList))
+	assert.Nil(t, verifyValidatorSetEquiv(vals, valsc))
+
+	// Verify a duplicate address across removes and adds is caught and vals is not modified
+	v2d := newValidator([]byte("v2"), 0)
+	v2m := newValidator([]byte("v2"), 10)
+	valList = []*Validator{v2d, v2m}
+	valsc = vals.Copy()
+	assert.NotNil(t, vals.UpdateWithChangeSet(valList))
+	assert.Nil(t, verifyValidatorSetEquiv(vals, valsc))
+
+	// Verify negative voting power is caught and vals is not modified
+	v1 = newValidator([]byte("v3"), -10)
+	valList = []*Validator{v1}
+	valsc = vals.Copy()
+	assert.NotNil(t, vals.UpdateWithChangeSet(valList))
+	assert.Nil(t, verifyValidatorSetEquiv(vals, valsc))
+
+	// Verify deleting unexisting validator is caught and vals is not modified
+	v1 = newValidator([]byte("v7"), 0)
+	valList = []*Validator{v1}
+	valsc = vals.Copy()
+	assert.NotNil(t, vals.UpdateWithChangeSet(valList))
+	assert.Nil(t, verifyValidatorSetEquiv(vals, valsc))
+
+	// Change priorities
+	v1 = newValidator([]byte("v1"), 10)
+	v2 = newValidator([]byte("v2"), 10)
+	valList = []*Validator{v1, v2}
+	assert.Nil(t, vals.UpdateWithChangeSet(valList))
+	assert.Nil(t, verifyValidatorSet(vals))
+
+	// Add new validators
+	v3 := newValidator([]byte("v3"), 200)
+	v4 := newValidator([]byte("v4"), 200)
+	newVals := []*Validator{v3, v4}
+	assert.Nil(t, vals.UpdateWithChangeSet(newVals))
+	assert.Nil(t, verifyValidatorSet(vals))
+
+	// Delete validators
+	v3 = newValidator([]byte("v3"), 0)
+	v4 = newValidator([]byte("v4"), 0)
+	delVals := []*Validator{v3, v4}
+	assert.Nil(t, vals.UpdateWithChangeSet(delVals))
+	assert.Nil(t, verifyValidatorSet(vals))
```

Comment:
This looks like this could be changed to [table-driven tests](https://github.com/golang/go/wiki/TableDrivenTests).

Reply:
Ok, let me try this.

Reply:
changed as suggested